src/Query/IntermediateTaskGroups.ts

```diff
@@ -11,6 +11,67 @@ import { Group } from './Group';
  */
 export class IntermediateTaskGroupsStorage extends Map<string[], Task[]> {}
 
+/**
+ * A node in the grouping tree. The node contains the
+ * list of tasks matching the path from the root so far, and its children
+ * are the further grouping of those tasks.
+ *
+ * Check the docs of @GroupingTree for more details.
+ *
+ */
+class GroupingTreeNode {
+    children: Map<string, GroupingTreeNode> = new Map();
+    tasks: Task[] = [];
+
+    /**
+     * Recursively traverse the tree to generate all the paths to the leaves.
+     * This function returns a map from every leaf path, to the list of tasks
+     * matching this path.
+     */
+    generateAllPaths(pathSoFar: string[] = []): IntermediateTaskGroupsStorage {
+        const resultMap = new IntermediateTaskGroupsStorage();
+        if (this.children.size == 0) {
+            // Base case: Leaf node. Populate the results map with the path to
+            // this node, and the tasks that match this path.
+            resultMap.set([...pathSoFar], this.tasks);
+            return resultMap;
+        }
+
+        for (const [property, child] of this.children) {
+            pathSoFar.push(property);
+            const childResult = child.generateAllPaths(pathSoFar);
+            childResult.forEach((value, key) => resultMap.set(key, value));
+            pathSoFar.pop();
+        }
+        return resultMap;
```

Comment:
I think there is some important logic in here. If maintaining it, I would have to concentrate quite hard to be confident of not breaking the code.

Looking at the new tests, I wonder how many corner cases in this code are explicitly being tested?

The original author of this plugin, when reviewing the original Grouping code, asked for each non-trivial class to generally be in its own file.

So I wonder about separating this out in to it own class, and have some tests on its behaviour?

Reply:
The tests you added are really clear, and give a good idea of what the code does. Thank you!